mne/simulation/source.py

```diff
@@ -312,3 +313,111 @@ def simulate_stc(src, labels, stc_data, tmin, tstep, value_fun=None,
     stc = SourceEstimate(np.concatenate(data), vertices=vertno, tmin=tmin,
                          tstep=tstep, subject=subject)
     return stc
+
+
+class SourceSimulator():
+    """
+    Simulate Stcs
+    """
+
+    def __init__(self, src, tmin=None, tstep=None, subject=None, verbose=None,
+                 duration=None):
+        self.src = src
+        self.tmin = tmin
+        self.tstep = tstep
+        self.subject = subject
+        self.verbose = verbose
+        self.labels = []
+        self.waveforms = []
+        self.events = np.empty((0, 3))
+        self.duration = duration
+        self.last_sample = []
+
+    def add_data(self, source_label, waveform, events):
+        '''
+        '''
+        # Check for mistakes
+        # Source_labels is a Labels instance
+        if not isinstance(source_label, Label):
+            raise ValueError('source_label must be a Label,'
+                             'not %s' % type(source_label))
+        # Waveform is a np.array or list of np arrays
+        # If it is not a list then make it one
+        if not isinstance(waveform, list) or len(waveform) == 1:
+            waveform = [waveform]*len(events)
+            # The length is either equal to the length of events, or 1
+        if len(waveform) != len(events):
+            raise ValueError('Number of waveforms and events should match'
+                             'or there should be a single waveform')
+        # Update the maximum duration possible based on the events
+        # imax = np.argmax(events[:,2])
+        # if events[imax,2]+len(waveform)
+        self.labels.extend([source_label]*len(events))
+        self.waveforms.extend(waveform)
+        self.events = np.vstack([self.events, events])
```

Comment:
Not essential, but I think a lot of the code would be streamlined if events where split to have 1 waveform, 1 label, 1 event.